ibmdbpy/base.py

```diff
@@ -188,28 +188,45 @@ def __init__(self, dsn, uid='', pwd='', autocommit=True, verbose=False):
 
         if self._con_type == 'jdbc':
 
+            missingCredentialsMsg = ("Missing credentials to connect via JDBC.")
+            ambiguousDefinitionMsg = ("Ambiguous definition of userID or password: " +
+                                      "Cannot be defined in uid and pwd parameters " +
+                                      "and in jdbc_url_string at the same time.")
+
             # remove trailing ":" or ";" or spaces on dsn, we will replace.
             dsn = dsn.rstrip(';: ')
             # find parameters on dsn; if any exist, there will be an equals sign.
             ix = dsn.find("=")
             # if no parameters exist, then this is the complete dsn
             if (ix < 0):
-                # confirm uid and pwd args are present, do NOT allow blank password
-                if (uid == '') | (pwd == ''):
-                    message = ("Missing credentials to connect via JDBC.")
-                    raise IdaDataBaseError(message)
-                # add uid, pwd parameters.
-                dsn = dsn + ':user={};password={};'.format(uid, pwd)
+                # nothing needs to be done, if uid and pwd are missing
+                # (uid and pwd are not needed for local database connections)
+                # otherwise both uid and pwd have to be specified
+                if (uid != '') | (pwd != ''):
+                    # confirm both uid and pwd args are present
+                    if (uid == '') | (pwd == ''):
+                        raise IdaDataBaseError(missingCredentialsMsg)
+                    else:
+                        # add uid, pwd parameters.
+                        dsn = dsn + ':user={};password={};'.format(uid, pwd)
             else:
                 # if we know there is at least one parameter; we can assume there exists a ":" before the parameter
                 # portion of the string in a correctly formatted dsn.  Therefore, just check for the existence of 
-                # the uid and pwd and add if they are missing.  If they are on the string, IGNORE these arguments
-                # and keep the string as-is.
+                # the uid and pwd and add if they are missing.  If they are on the string, IGNORE eep the string as-is.
                 if not ('user=' in dsn):
+                    if (uid == ''):
```

Comment:
This assumes that if there's an `=` in the connection string, either `user=` has to be in the connection string or `uid` has to be supplied. However, what about local connections that don't require authentication information where the user still wants to set some properties on the connection string? I admit that I couldn't find information in the knowledge center that confirms or denies whether that's possible.

Reply:
Good point! Your assumption is correct, you can specify properties for a local connection, for instance `jdbc:db2:BLUDB:specialRegisters=CURRENT SCHEMA=TEST;`.
I'll make the necessary changes in the code.

Reply:
I have made the necessary changes with commit https://github.com/ibmdbanalytics/ibmdbpy/pull/59/commits/d78c4793d7850d88d4ca8508419588151ef662f3
and the FVT testcases completed successfully:
 https://hyc-idax-jenkins.swg-devops.com/view/Custom%20FVT/job/FVT_CUSTOM_IBMDBPY/#20/